Add files via upload (#2215)
elastic/ecs: rfcs/0000-User&HostIntegrationRFC.md

# 0000: Asset Integration
<!-- Leave this ID at 0000. The ECS team will assign a unique, contiguous RFC number upon merging the initial stage of this RFC. -->

- Stage: **0 (strawperson)** <!-- Update to reflect target stage. See https://elastic.github.io/ecs/stages.html -->
- Date: **TBD** <!-- The ECS team sets this date at merge time. This is the date of the latest stage advancement. -->

<!--
As you work on your RFC, use the "Stage N" comments to guide you in what you should focus on, for the stage you're targeting.
Feel free to remove these comments as you go along.
-->

<!--
Stage 0: Provide a high level summary of the premise of these changes. Briefly describe the nature, purpose, and impact of the changes. ~2-5 sentences.
-->

This proposal aims to extend the existing ECS fieldset to store inventory metadata for hosts, and users from external application repositories. Using ECS to store such fields will improve metadata querying and retrieval across various use cases.

> Within the Security `Entity Analytics` initiative, we refer to hosts and users as `entities`. We refer to hosts/ users as `assets` across generic security or observability use cases. For simplification, entities, and assets will refer to host and user objects.

Author: 👋 Andrew, I need clarification, please. Are you suggesting we adopt the 'device' terminology in this RFC scope definition/  ECS schema definition/ for the asset management integration use case (a subset of this RFC)/ all of the above? AFAIK, we have not yet adopted the device terminology anywhere in our solution yet. Thanks.

@lauravoicu curious to hear your thoughts from an operational perspective. We have a mixed bag of vendor terminologies to reconcile from.

> Within the Security `Entity Analytics` initiative, we refer to hosts and users as `entities`. We refer to hosts/ users as `assets` across generic security or observability use cases. For simplification, entities, and assets will refer to host and user objects.

Author: @andrewkroh I'm adding additional clarifications to this section for now.

-->

### Proposed New Fields for User object

Field | Type | Example | Description
--- | --- | --- | ---
user.profile.id	| keyword	| 1234	| User ID from the identity datasource.
user.profile.type	| keyword	| Employee	| Type of user account.
user.profile.status	| keyword |	On board	| Status of the user account.
user.profile.first_name	| keyword |	First	| First Name of the User.
user.profile.last_name	| keyword |	Last	| Last Name of the user.
user.profile.other_identities	| keyword , multifield |	first.last@elk.elastic.co	| Array of additional user identities.
user.profile.manager	| keyword |	John Doe	| Assigned Manager for the user account.

> user.profile.employee_type	| keyword |	Regular | Further classification type for the user account.

Reviewer: Capturing this level of user profile data as part of an observability system is extremely problematic due to the PII nature of this data.

The `user.profile.*` listed here is more like the information that would be stored (and protected) by the identity system and only available on the backend and as part of the encrypted details in any token.

i.e. They should not be included with general observability data.

Reviewer: I would then recommend that a note (at the least) be added to these fields. Identifying there is possible PII issues that sending / storing this data might cause.

e.g. sending / storing this may lead to EUDB violations for example (depending on lots of caveats and how the data is store / whether it can be removed (right to forget) etc.).

 For some people / use cases this might be an acceptable option, but we don't really want to have anyone just including the data because they can.

Author: 👋 @MSNev, thanks for taking the time to review. Great observations.

This subset of the user.* may be most prudent for security use cases and likely not in general observability use cases. Depending on the identity application mappings and definitions, these fields may also contain PII data.

I proposed these to users.* schema changes for three primary reasons:
1. This security use case of ingesting Identity metadata assumes appropriate index-level access within ElasticSearch and by the security solution.
2. ECS does not restrict whether a field should/should not be populated with PII. 
3. Theoretically, if you allow me to generalize, the data mapping definitions are abstracted from how the data is consumed within an application (viewed, queried, etc.)

We have previously discussed general approaches for pseudonymizing data within the ECS framework. Thus, I recommend standardizing these mapping definitions and expanding the data anonymization concepts in parallel. 

Please lmk if these thoughts resonate with you or if further clarifications are needed.

user.profile.job_family	| keyword |	65-Sales	| Job family associated with the user account.
user.profile.job_family_group	| keyword |	GTM	| Job family group associated with the user account.
user.profile.management_level	| keyword |	Individual Contributor	| If the user account is identified as a Manager or Individual contributor.
user.profile.job_title	| keyword |	Field Sales	| Job title assigned to the user account.
user.profile.department	| keyword |	x256	| Department name associated with the user account.
user.profile.organization	| keyword |	Elasticsearch Inc.	| Organization name associated with the account.
user.profile.location	| keyword |	US - Washington - Distributed	| Assigned location for the user account.
user.profile.mobile_phone	| keyword |	222-222-2222	
user.profile.primaryPhone	| keyword |	222-222-2222	
user.profile.secondEmail	| keyword |	first.l@elastic.co	| Additional email addresses associated with the user account.
user.profile.sup_org_id	| keyword |	SUP-ORG-75	| Primary organization ID for the user account.
user.profile.supervisory_Org	| keyword |	Field Sales	| Primary organization name for the user account.

> user.profile.assigned_mdm_id	| keyword |	2950	| Primary host owned by the user. This acts as a correlation to identify the host event.			

Reviewer 2: So this correlates to an `asset.id` value? If so, I think the description should explicitly state this.

How do we qualify what is the "primary" device? For example, a user might be associated with a laptop, tablet, and phone.

Author: It can be the `asset.id` field value. It depends on the user's implementation of directory services and how they map their identifiers.

This ECS scope does not enforce or validate these correlation values. The data query logic will vary by the use case and is abstracted from this ECS definition.

I clarified the same in the field description- line 67.

> user.profile.assigned_mdm_id	| keyword |	2950	| Primary host owned by the user. This acts as a correlation to identify the host event.			

Reviewer 2: Should there be a way to associate the user to more than once device? I'm not saying there should be, because if they are not explicitly "linked" the association can still be made by querying for `asset.assigned_users: {user.profile.id}`.

user.account.isDelegatedAdmin	| boolean |	true/ false	| A flag indicating if account has Delegated Admin rights
user.account.isPriviledged	| boolean |	true/ false	| A flag indicating if account is a Privileged account
user.account.status.password_expired	| boolean |	true/ false	| A flag indicating if account password has expired.
user.account.status.deprovisioned	| boolean |	true/ false	| A flag indicating if account has been deprovisioned
user.account.password_change_date	| date |	June 5, 2023 @ 18:25:57.000	| Last date/time when account password was updated

### Proposed New Fields for Asset object

Field | Type | Generic Example |	User Entity Example | Host Entity Example | Description
--- | --- | --- | --- | --- | ---
asset.category	| keyword |	-	        | Null	                | hardware	                | A further classification of the asset type beyond event.category. For example for host assets {hardware, virtual, container, node}. For user assets {NULL ?}			
asset.type	    | keyword |	-	        | Null	                | workstation	            | A sub classification of asset. For host assets {workstation, S3, Compute}. For user assets {NULL?}.			

> asset.id	    | keyword |	-	        | 00uhs72c27s6PiK7x1t7	| 2950	                    | A unique ID for the asset. For inventory integrations, it's the id generated from inventory datasource			

Author: We recommend the `asset ID` be unique within a user's environment. For external directory services, this can map to the directory asset id. 

It will largely depend on the source and management of the data (or how a user configures their external asset management or directory service application).

For the o11y use case you mention above, can this map to the asset.ean field and ensure uniqueness?

> asset.id	    | keyword |	-	        | 00uhs72c27s6PiK7x1t7	| 2950	                    | A unique ID for the asset. For inventory integrations, it's the id generated from inventory datasource			

Reviewer 3: What we are currently targeting as a model in observability is to use `asset.id` to represent the "internal" ID of that asset, so for example, for a container asset, that container's "Container ID" would be the "asset.id" value. Within all container assets, we expect that to be unique. Pod assets would list their pod ID there, as well. Hosts might use something like an FQDN hostname value there. In some cases, this value may be reasonably unique within the set of assets of that single type, but not unique across assets of other types. For that reason, we have the `asset.ean` value, which at the moment concatenates `asset.kind` + `asset.id` to help namespace the internal asset ID values. The EAN also allows us to expand the convention, URN-style, to include more information in the overall asset's identifier, while still maintaining that asset's own concept of its own ID within its asset space (e.g. container ID).

asset.create_date	            | date |	June 5, 2023 @ 18:25:57.001	| - | -             | For users, it's the hire date, for other assets, it's the in service date.			
asset.end_date	                | date |	June 5, 2023 @ 18:25:57.002	| - | -             | For users, it's the termination date, for other assets, it's the out-of-service date.			
asset.first_seen	            | date |	June 5, 2023 @ 18:25:57.003	| - | -             | The earliest date/time at which this asset was observed.  For assets added via config, this would remain empty until the asset was observed.			
asset.last_seen	                | date |	June 5, 2023 @ 18:25:57.004	| - | -             | The most recent date/time at which this asset was observed.  Would remain empty until the asset was observed.			
asset.last_updated	            | date |	June 5, 2023 @ 18:25:57.005	| - | -             | The most recent date/time this asset was updated in the inventory datasource.		
asset.serial_number	            | keyword	| C02FG1G1MD6T	| - | -             |		Serial number of the asset.
asset.tags	                    | keyword	  | watch, mdmaccess		| - | -             |	Tags assigned at the MDM.

> asset.assigned_users	          | keyword	  | user1@email.com, user2@email.com		| - | -             |	List of users assigned to the asset.

Reviewer 2: I think the value of the `asset.owner` should always be included in this list. This will make is simpler to query for devices associated to a user because only one field needs to be searched. If you agree, then can you explicitly state in the description that the `asset.owner` value should always be included.

> asset.assigned_users	          | keyword	  | user1@email.com, user2@email.com		| - | -             |	List of users assigned to the asset.

Author: I have incorporated your inputs in lines 52, 93, and 108.

# 0000: Asset Integration
<!-- Leave this ID at 0000. The ECS team will assign a unique, contiguous RFC number upon merging the initial stage of this RFC. -->

- Stage: **0 (strawperson)** <!-- Update to reflect target stage. See https://elastic.github.io/ecs/stages.html -->
- Date: **TBD** <!-- The ECS team sets this date at merge time. This is the date of the latest stage advancement. -->

<!--
As you work on your RFC, use the "Stage N" comments to guide you in what you should focus on, for the stage you're targeting.
Feel free to remove these comments as you go along.
-->

<!--
Stage 0: Provide a high level summary of the premise of these changes. Briefly describe the nature, purpose, and impact of the changes. ~2-5 sentences.
-->

> This proposal aims to extend existing ECS fieldset to store inventory metadata for hosts, users from external application repositories. Using ECS to store such fields will improve metadata queryring and retrieval across various use cases.

Reviewer 4: ```suggestion
This proposal aims to extend existing ECS fieldset to store inventory metadata for hosts, users from external application repositories. Using ECS to store such fields will improve metadata querying and retrieval across various use cases.
```

Within the Security `Entity Analytics` initiatve, we refer to hosts and users as `entities`. We also refer to hosts/ users as `assets` across generic security or obserbality use cases. For simplificaiton, entities, assets will refer to host and user objects.

This proposal includes the following:
* Additional fields in the `users` object 
* Introduces a new field set called `assets`
<!-- * Additional fields in the `host` object --->

This proposal will also facilitate storing of host and user inventory within security solution (aka. the entity store).

### References:
* [Entity Analytics Plugin Architecture for Host & User Integrations](https://github.com/elastic/beats/issues/33866#issuecomment-1165709940)
* 0038: Extend event.kind allowed values with asset [Existing RFC](https://github.com/elastic/ecs/blob/main/rfcs/text/0038-event-kind-asset.md#0038-extend-eventkind-allowed-values-with-asset)

> * [Creation of an asset `super-schema`](https://github.com/elastic/security-team/issues/2347)

Reviewer 4: Please remove links to non-public repos since this proposal will be public facing. If you want to link the issues together, comment on the private issue referencing the public one.

<!--
Stage 1: If the changes include field additions or modifications, please create a folder titled as the RFC number under rfcs/text/. This will be where proposed schema changes as standalone YAML files or extended example mappings and larger source documents will go as the RFC is iterated upon.
-->

<!--
Stage X: Provide a brief explanation of why the proposal is being marked as abandoned. This is useful context for anyone revisiting this proposal or considering similar changes later on.
-->

## Fields

<!--
Stage 1: Describe at a high level how this change affects fields. Include new or updated yml field definitions for all of the essential fields in this draft. While not exhaustive, the fields documented here should be comprehensive enough to deeply evaluate the technical considerations of this change. The goal here is to validate the technical details for all essential fields and to provide a basis for adding experimental field definitions to the schema. Use GitHub code blocks with yml syntax formatting, and add them to the corresponding RFC folder.
-->

### Proposed New Fields for User object

Field | Type | Example | Description
--- | --- | --- | ---

> user.profile.id	| keyword	| 1234	| User ID from the identity datasource.

Reviewer 4: What is the significance for introducing `user.profile.id` vs. using the existing `user.id` field?

Author: I am just getting this resolved for now. Please lmk, or reopen, if we need to reconsider.

user.profile.type	| keyword	| Employee	| Type of user account.
user.profile.status	| keyword |	On board	| Status of the user account.
user.profile.first_name	| keyword |	First	| First Name of the User.
user.profile.last_name	| keyword |	Last	| Last Name of the user.

> user.profile.other_identities	| keyword , multifield |	first.last@elk.elastic.co	| Array of additional user identities.

Reviewer 4: What type of `multifield` would be used here? `.text`?

Reviewer 4: Can that be captured in the doc?

user.profile.manager	| keyword |	John Doe	| Assigned Manager for the user account.
user.profile.employee_type	| keyword |	Regular | Further classification type for the user account.
user.profile.job_family	| keyword |	65-Sales	| Job family associated with the user account.
user.profile.job_family_group	| keyword |	GTM	| Job family group associated with the user account.
user.profile.management_level	| keyword |	Individual Contributor	| If the user account is identified as a Manager or Individual contributor.
user.profile.job_title	| keyword |	Field Sales	| Job title assigned to the user account.
user.profile.department	| keyword |	x256	| Department name associated with the user account.
user.profile.organization	| keyword |	Elasticsearch Inc.	| Organization name associated with the account.
user.profile.location	| keyword |	US - Washington - Distributed	| Assigned location for the user account.

> user.profile.mobilePhone	| keyword |	222-222-2222	

Reviewer 4: ```suggestion
user.profile.mobile_phone	| keyword |	222-222-2222	
```

ECS fields [conventions](https://www.elastic.co/guide/en/ecs/current/ecs-guidelines.html#_guidelines_for_field_names) use all lower case letters and underscore (`_`) to combine words in field names

asset.tags	                    | keyword	  | watch, mdmaccess		| - | -             |	Tags assigned at the MDM.
asset.assigned_users	          | keyword	  | user1@email.com, user2@email.com		| - | -             |	List of users assigned to the asset.
asset.assigned_users_are_admin	| boolean	  | TRUE	| - | -             |		Flag to identify if the assigned users have admin priviledges.
asset.is_managed	              | boolean	  | TRUE			| - | -             | If asset is managed by the organization.
asset.last_enrolled_date	      | date	    | June 5, 2023 @ 18:25:57.005		| - | -             |	The most recent date/time the asset checked in with MDM.
asset.data_classification	      | keyword	  | restricted		| - | -             |	Data classification tier for the asset.

#### Nesting of existing risk.* fields under asset object
* We have a set of risk.* fields in ECS which can be further nested under the asset.* object. Reference to [Risk RFC](https://github.com/elastic/ecs/blob/main/rfcs/text/0031-risk-fields.md).

### Proposed New Fields for Host object
Field | Type | Example | Description
--- | --- | --- | ---

> host.os.build	| keyword		| 22F66   | Host OS Build information

Reviewer 4: The current convention for build info is populating into the `host.os.kernel` (e.g. `winlogbeat` populates with `10.0.17763.4377 (WinBuild.160101.0800)` on a Windows Server 2019 Datacenter system). Is the idea here to create another field solely for the build number?

For this change as proposed, the field would go in as `os.build` under `os.*` and populate anywhere `os.*` fields are nested: `host.os.build`, `observer.os.build`, and `user_agent.os.build`.

Reviewer 4: > Is the idea here to create another field solely for the build number?

Let's add this under `Concerns`. Adding and populating an `os.build` field isn't breaking, but it is a notably change from the current convention.

Author: Addressed this as part of this commit - https://github.com/elastic/ecs/pull/2215/commits/fd289d1eac3c22a64e80d18989d64443ca8a6814

> host.installed_extensions 	| keyword	  | Nested objects	  | List of installed extensions along with their metadata

Reviewer 4: What is the motivation for adding `installed_extensions` and `installed_applications` underneath `host.*` instead of the proposed `asset.*` field set?

Reviewer 4: I'm against about crowding the existing `host.*` field set if the main use for these new fields is tied to asset and entity store use cases. The `host.*` fields are mostly populated by the [add_host_metadata](https://www.elastic.co/guide/en/beats/filebeat/current/add-host-metadata.html) processor for each event, and we wouldn't want to add the list of `installed_extensions` and `installed_applications` to every produced event from the host.

In my own experience, while VM/container instances are tied to an underlying physical host, the instances are also treated as individual compute assets: unique owners, cost center, business owners, admins.

Author: > The `host.*` fields are mostly populated by the [add_host_metadata](https://www.elastic.co/guide/en/beats/filebeat/current/add-host-metadata.html) processor for each event, and we wouldn't want to add the list of `installed_extensions` and `installed_applications` to every produced event from the host.

The reference and elaboration are helpful in understanding the complexity. Let's move these two fields under assets.*.

Author: I addressed this as part of this commit - https://github.com/elastic/ecs/pull/2215/commits/fd289d1eac3c22a64e80d18989d64443ca8a6814

host.installed_applications	| keyword	    | Nested objects	  | List of installed applications along with their metadata

<!--
Stage 2: Add or update all remaining field definitions. The list should now be exhaustive. The goal here is to validate the technical details of all remaining fields and to provide a basis for releasing these field definitions as beta in the schema. Use GitHub code blocks with yml syntax formatting, and add them to the corresponding RFC folder.
-->

## Usage

<!--
Stage 1: Describe at a high-level how these field changes will be used in practice. Real world examples are encouraged. The goal here is to understand how people would leverage these fields to gain insights or solve problems. ~1-3 paragraphs.
-->

> * As part of Entity Analytics, we are ingesting metadata about [Users](https://github.com/elastic/security-team/issues/4163) and [Hosts](https://github.com/elastic/security-team/issues/4044) from various external vendor applications. We are storing all ingested metadat in ElasticSearch inde. After we map these fields to ECS, we will enrich these ingested events for risk scoring scenarios (e.g. of context enrichments) and detecting advanced analytics (UBA) use cases. 

Reviewer 4: ```suggestion
* As part of Entity Analytics, we are ingesting metadata about [Users](https://github.com/elastic/security-team/issues/4163) and [Hosts](https://github.com/elastic/security-team/issues/4044) from various external vendor applications. We are storing all ingested metadata in Elasticsearch index. After we map these fields to ECS, we will enrich these ingested events for risk scoring scenarios (e.g. of context enrichments) and detecting advanced analytics (UBA) use cases. 
```

Please remove links to non-public issues.

* This schema will be used to persist `Observed` (queried) entities from ingested security log dataset in an Entity store. This entity store can be further extended to meet broaded overall Asset Mangagement needs.

* We also envision additional enrichment use cases for existing prebuilt detection rules down the line.

## Source data

<!--
Stage 1: Provide a high-level description of example sources of data. This does not yet need to be a concrete example of a source document, but instead can simply describe a potential source (e.g. nginx access log). This will ultimately be fleshed out to include literal source examples in a future stage. The goal here is to identify practical sources for these fields in the real world. ~1-3 sentences or unordered list.
-->

There are many sources of asset inventory repositories. In the mid-term we are planning to ingest data from the following applicaiton providers:

### User (Identity) repository sources:

> *  Azure Active Directory

Reviewer 4: ```suggestion
* Azure Active Directory
```

nit: removing extra spacing

* Active Directory DS
* Okta 
* Workday
* GSuite 
* GitHub

### Host repository sources:
* Azure Active Directory
* Jamf 
* Active Directory DS
* MS Intune
* ServiceNow Asset CMDB

> Mapping of application fields from individual providers are documented here: [User](https://docs.google.com/spreadsheets/d/1ZHGQPGVr35vLZB3DcPnXWxLTkHad_S0YrhtTKO_LL4o/edit#gid=1940810377), [Host-TBD]().

Reviewer 4: Like any linked GitHub issues, any other linked resources should be publicly viewable for the community.

-->

* We have a couple fleet integrations under development. We want them to use these proposed ECS before being released.
* Schema/ field sets defined here focuses on asset inventory datasources. Additional fields may need to be appeneded (ideally within this RFC lifecyle) to support the entity store needs.
* Due deligence is needed to avoid proliferation of field sets and validating business requiurements.
* ?

<!--
Stage 2: Document new concerns or resolutions to previously listed concerns. It's not critical that all concerns have resolutions at this point, but it would be helpful if resolutions were taking shape for the most significant concerns.
-->

<!--
Stage 3: Document resolutions for all existing concerns. Any new concerns should be documented along with their resolution. The goal here is to eliminate risk of churn and instability by ensuring all concerns have been addressed.
-->

## People

The following are the people that consulted on the contents of this RFC.

* @sourinpaul | author
* @andrewkroh | subject matter expert

> * @jamie Hynds | subject matter expert

Reviewer 4: ```suggestion
* @jamiehynds | subject matter expert
```

nit: fixing handle

* @lauravoicu | subject matter expert
* @MikePaquette | subject matter expert
* @sourinpaul | sponsor
* ?

<!--
Who will be or has been consulted on the contents of this RFC? Identify authorship and sponsorship, and optionally identify the nature of involvement of others. Link to GitHub aliases where possible. This list will likely change or grow stage after stage.

e.g.:

* @Yasmina | author
* @Monique | sponsor
* @EunJung | subject matter expert
* @JaneDoe | grammar, spelling, prose
* @Mariana
-->

## References

<!-- Insert any links appropriate to this RFC in this section. -->

### RFC Pull Requests

<!-- An RFC should link to the PRs for each of it stage advancements. -->

> * Stage 0: https://github.com/elastic/ecs/pull/NNN

Reviewer 4: ```suggestion
* Stage 0: https://github.com/elastic/ecs/pull/2215
```
